Add Galen's Blog Post and Update Blog Image Hosting (#234)
ComputerScienceHouse/CSHPublicSite: _posts/2020-05-19-docker-rust-notpresent.md

---
title: "Debugging Error: NotPresent with Rust in Docker"
description: "A friend's Docker container was exiting with Error: NotPresent and no other information. I embarked on a deep dive for an ultimately simple cause."
layout: post
date: 2020-05-19
image: https://assets.csh.rit.edu/pubsite/projects/carbon.png
image-sm: https://assets.csh.rit.edu/pubsite/projects/carbon.png
author: Galen Guyer
author-image: https://cdn.galenguyer.com/img/avatars/headshot.jpg
author-email: chef@csh.rit.edu
author-social:
    github: https://github.com/galenguyer/
categories: [docker, rust]
---

<a href='https://galenguyer.com/blog/2020/05/19/docker-rust-notpresent'>Check out this post on Galen's blog!</a>

<p>Today, a friend posted the following snippet in Slack:</p>
<pre> [~/Documents/shelflife] [±docker✗] $ docker run -v -it --rm --name my-running-app shelflife-rust-docker
    Error: NotPresent</pre>

> <p>The first guess was they deleted the docker image and forgot to rebuild it. That wasn't the case, leading me to dive into this cryptic error message that we assumed some some hidden Rust or Docker error. We weren't able to find any example of this error message for Rust or Docker online, so I knew this would be a fun challenge.</p>

Reviewer: some some should be was some

> <p>The first guess was they deleted the docker image and forgot to rebuild it. That wasn't the case, leading me to dive into this cryptic error message that we assumed some some hidden Rust or Docker error. We weren't able to find any example of this error message for Rust or Docker online, so I knew this would be a fun challenge.</p>

Reviewer 2: ```suggestion
<p>The first guess was that they had deleted the docker image and forgot to rebuild it. That wasn't the case, leading me to dive into this cryptic error message that we assumed some some hidden Rust or Docker error. We weren't able to find any example of this error message for Rust or Docker online, so I knew this would be a fun challenge.</p>
```

> <p>The first thing I did was to make sure the path to the binary was working, it'd be silly if this wasn't working because <code>/usr/local/bin</code> wasn't in the container's PATH. I changed the last line in the Dockerfile from <code>CMD ["shelflife"]</code> to <code>ENTRYPOINT ["/usr/local/bin/shelflife"]</code>. The error persisted, which means it wasn't a PATH error. Of course this couldn't be easy.</p>

Reviewer: add "as" before it'd be silly

> <p>The first thing I did was to make sure the path to the binary was working, it'd be silly if this wasn't working because <code>/usr/local/bin</code> wasn't in the container's PATH. I changed the last line in the Dockerfile from <code>CMD ["shelflife"]</code> to <code>ENTRYPOINT ["/usr/local/bin/shelflife"]</code>. The error persisted, which means it wasn't a PATH error. Of course this couldn't be easy.</p>

Reviewer 2: ```suggestion
<p>The first thing I did was to make sure the path to the binary was working, it'd be silly if this wasn't working because <code>/usr/local/bin</code> wasn't in the container's PATH. I changed the last line in the Dockerfile from <code>CMD ["shelflife"]</code> to <code>ENTRYPOINT ["/usr/local/bin/shelflife"]</code>. The error persisted, which meant it wasn't a PATH error. Of course this couldn't be easy.</p>
```

> <p>The next step in diagnosis was to get a shell on the container. That'd typically be done via <code>docker exec</code>, but that only works on a running container, which we didn't have. To get around this, I changed the last line of the Dockerfile to <code>ENTRYPOINT ["/bin/bash"]</code>, which makes <code>docker run -it</code> drop you into an interactive bash shell. This means I can easily see what's going on and run commands from inside the container.</p>

Reviewer: can easily -> could easily

> <p>Now that I had a shell, I navigated to <code>/usr/local/bin/</code> and ran <code>./shelflife</code> to check if there was any output docker was hiding. Unfortunately, there wasn't. I was at roughly the same place as when I started - a cryptic error message and no idea where it was coming from. At least now we can be decently sure it's not a docker issue.</p>

Reviewer: can be decently sure -> was decently sure

> <p>I'd ran into issues with dynamically linked libraries not being installed in the past, so I used <code>ldd</code> (List Dynamic Dependencies) to make sure all the dependencies existed. Everything looked fine, and I double-checked by manually installing the library packages for everythin <code>ldd</code> said was linked. Everything was already installed, and the error persisted.</p>

Reviewer 2: ```suggestion
<p>I'd run into issues with dynamically linked libraries not being installed in the past, so I used <code>ldd</code> (List Dynamic Dependencies) to make sure all the dependencies existed. Everything looked fine, and I double-checked by manually installing the library packages for everything <code>ldd</code> said was linked. Everything was already installed, but the error persisted.</p>
```
